Modifying builtins-c-common.hpp does not trigger a rebuild

When `builtins-c-common.hpp` is modified, I expect the depending files to be recompiled. For example, `builtins-c-cpu.hpp` should be compiled again. However, this is not the case: the modification is silently ignored. This results in sneaky issues.

**EDIT:**
There is an even bigger issue: when the cpp files cannot be compiled due to errors in it, errors are shown (and ignored) but it does not break the build the first time. In subsequent builds, the file is not even re-compiled! It looks like an empty llvm file is generated in this case and this is used to generate a bc file which actually contains nothing... This certainly results in weird bug later (possibly even at runtime).

> **EDIT:** There is an even bigger issue: when the cpp files cannot be compiled due to errors in it, errors are shown (and ignored) but it does not break the build the first time. In subsequent builds, the file is not even re-compiled! It looks like an empty llvm file is generated in this case and this is used to generate a bc file which actually contains nothing... This certainly results in weird bug later (possibly even at runtime).

You mean builtins-c-cpu.cpp, right? There's one strange behavior that I'm not sure is intentional. Our current build system silently fails compilation of these files when generating bc files for certain arch/os combinations. For example, when the required dependencies aren't installed in the build environment. This results in that configuration being skipped from the supported matrix.

While this is technically a bug in our build system, we kind of rely on it to let builds succeed in this "fuzzy" manner across various environments. Fixing it would break some build setups, such as this one in our CI job: [link](https://github.com/ispc/ispc/actions/runs/25613259232/job/75186921125?pr=3830): 

```bash
[ 14%] Generating share/ispc/builtins_cpp_32_linux_armv8a.bc
In file included from builtins/builtins-c-cpu.cpp:54:
In file included from /usr/include/stdio.h:444:
/usr/include/bits/floatn.h:83:52: error: unsupported machine mode '__TC__'
   83 | typedef _Complex float __cfloat128 __attribute__ ((__mode__ (__TC__)));
      |                                                    ^
/usr/include/bits/floatn.h:97:9: error: __float128 is not supported on this target
   97 | typedef __float128 _Float128;
      |         ^
2 errors generated.
make[2]: *** [CMakeFiles/common-builtins-bc.dir/build.make:84: share/ispc/builtins_cpp_32_linux_armv8a.bc] Error 1
make[1]: *** [CMakeFiles/Makefile2:484: CMakeFiles/common-builtins-bc.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 14%] Building CXX object CMakeFiles/common.dir/src/llvmutil.cpp.o
[ 14%] Built target generate-parser
```

We can certainly fix our own CI, but I'm less sure about all the users who build ISPC from source. I'm not convinced we need to fix this bug, but the first part about the dependency chain definitely needs to be fixed.

--- a/cmake/GenerateBuiltins.cmake
+++ b/cmake/GenerateBuiltins.cmake
@@ -63,6 +63,15 @@ if (${LLVM_VERSION_NUMBER} VERSION_GREATER_EQUAL "20.1.2")
         builtins/target-avx10_2-x64-common.ll)
 endif()
 
+# Headers included by builtins/builtins-c-cpu.cpp. Listed explicitly so that
+# add_custom_command rebuilds the bitcode when these headers change. DEPFILE
+# support in add_custom_command is generator-specific for CMake < 3.20, so we
+# follow the same pattern as M4_IMPLICIT_DEPENDENCIES above.
+list(APPEND BUILTINS_C_CPU_HEADER_DEPS
+    builtins/array.hpp
+    builtins/builtins-c-common.hpp
+    src/builtins-info.h)
+
 function(target_ll_to_cpp target bit os CPP_LIST BC_LIST)
     set(input builtins/target-${target}.ll)
     set(include builtins)
@@ -164,7 +173,7 @@ function(builtin_wasm_to_cpp bit os arch)
         OUTPUT ${bc}
         COMMAND ${EMCC_EXECUTABLE} ${flags} ${input} -o -
             | \"${LLVM_AS_EXECUTABLE}\" -o ${bc}
-        DEPENDS ${input}
+        DEPENDS ${input} ${BUILTINS_C_CPU_HEADER_DEPS}
         WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
     )
 
@@ -328,7 +337,7 @@ function(builtin_to_cpp bit os generic_arch)
         OUTPUT ${bc}
         COMMAND ${CLANGPP_EXECUTABLE} ${flags} ${input} -o -
             | \"${LLVM_AS_EXECUTABLE}\" -o ${bc}
-        DEPENDS ${input}
+        DEPENDS ${input} ${BUILTINS_C_CPU_HEADER_DEPS}
         WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
     )
 

